tests/src/analysis/testqgsprocessing.cpp

```diff
@@ -177,6 +177,220 @@ class DummyAlgorithm : public QgsProcessingAlgorithm
+      params.insert( "p4", true );
+      QVERIFY( !checkParameterValues( params, context, &message ) );
+      params.clear();
+      params.insert( "p3", true );
+      QVERIFY( checkParameterValues( params, context, &message ) );
+      params.insert( "p4", true );
+      QVERIFY( !checkParameterValues( params, context, &message ) );
+      params.remove( "p4" );
+      params.insert( "p2", true );
+      QVERIFY( !checkParameterValues( params, context, &message ) );
+      params.insert( "p1", true );
+      QVERIFY( checkParameterValues( params, context, &message ) );
+
+      // mutually_exclusive_required( all_required( p1, p2 ), all_required( p3, p4 ) )
+      mLogicalParameterGroups.parameters.clear();
+      mLogicalParameterGroups.childGroups.clear();
+      group.type = MutuallyExclusiveRequired;
+      group.parameters.clear();
+      group.childGroups.clear();
+
+      childGroup1.type = AllRequired;
+      childGroup1.parameters.insert( "p1" );
+      childGroup1.parameters.insert( "p2" );
+      childGroup2.type = AllRequired;
+      childGroup2.parameters.insert( "p3" );
+      childGroup2.parameters.insert( "p4" );
+
+      group.childGroups.append( childGroup1 );
+      group.childGroups.append( childGroup2 );
+
+      addLogicalParameterGroup( group );
+      params.clear();
+
+      QVERIFY( !checkParameterValues( params, context, &message ) );
+      params.insert( "p1", true );
+      QVERIFY( !checkParameterValues( params, context, &message ) );
+      params.insert( "p2", true );
+      QVERIFY( checkParameterValues( params, context, &message ) );
+      params.insert( "p3", true );
+      QVERIFY( !checkParameterValues( params, context, &message ) );
+      params.insert( "p4", true );
+      QVERIFY( !checkParameterValues( params, context, &message ) );
+      params.clear();
+      params.insert( "p3", true );
+      QVERIFY( !checkParameterValues( params, context, &message ) );
+      params.insert( "p4", true );
+      QVERIFY( checkParameterValues( params, context, &message ) );
+      params.insert( "p2", true );
+      QVERIFY( !checkParameterValues( params, context, &message ) );
+      params.insert( "p1", true );
+      QVERIFY( !checkParameterValues( params, context, &message ) );
+
+      // mutually_exclusive_optional( all_required( p1, p2 ) )
+      mLogicalParameterGroups.parameters.clear();
+      mLogicalParameterGroups.childGroups.clear();
+      group.type = MutuallyExclusiveOptional;
+      group.parameters.clear();
+      group.childGroups.clear();
+
+      childGroup1.type = AllRequired;
+      childGroup1.parameters.insert( "p1" );
+      childGroup1.parameters.insert( "p2" );
+      group.childGroups.append( childGroup1 );
+      addLogicalParameterGroup( group );
+      params.clear();
+
+      QVERIFY( checkParameterValues( params, context, &message ) );
+      params.insert( "p1", true );
+      QVERIFY( !checkParameterValues( params, context, &message ) );
```

Comment:
Is this test correct? for `mutually_exclusive_optional( all_required( p1, p2 ) )` I would expect that all four possible combinations are correct ({}, {p1}, {p2}, {p1, p2}).

Reply:
I think so - if you say both P1 AND P2 are required, then specifying only one should be a fail. You either need to leave both missing or specify both.

Reply:
But here in fact you are saying that both P1 AND P2 are required in a sub-term... And tehn the top-level `mutually_exclusive_optional( X )` is happy with both outcomes of `all_required( p1, p2 )` (satisfied or not satisfied) ?

Reply:
Yes - to me that's correct. It's saying you can optionally set the X and y resolution (the parent optional conditional), but if you specify one you HAVE to specify the other (the child all required condition)

Reply:
Yes - to me that's correct. It's saying you can optionally set the X and y resolution (the parent optional conditional), but if you specify one you HAVE to specify the other (the child all required condition)